build/print.go

```diff
@@ -631,60 +631,81 @@ const (
-		p.printf("%s", brack[:1])
-	}
-	p.depth++
-
-	// If there are line comments, force multiline
+// useCompactMode reports whether a sequence should be formatted in a compact mode
+func useCompactMode(start *Position, list *[]Expr, end *End, forceCompact, forceMultiLine bool) bool {
+	// If there are line comments, use multiline
 	// so we can print the comments before the closing bracket.
-	for _, x := range list {
+	for _, x := range *list {
 		if len(x.Comment().Before) > 0 {
-			forceMultiLine = true
+			return false
 		}
 	}
 	if end != nil && len(end.Before) > 0 {
-		forceMultiLine = true
+		return false
 	}
 
-	// Resolve possibly ambiguous call arguments explicitly
-	// instead of depending on implicit resolution in logic below.
-	if forceMultiLine {
-		forceCompact = false
+	// Implicit tuples are always compact
+	if !start.IsValid() {
+		return true
 	}
 
-	switch {
-	case len(list) == 0 && !forceMultiLine:
-		// Compact form: print nothing.
-
-	case len(list) == 1 && !forceMultiLine:
-		// Compact form.
-		p.expr(list[0], precLow)
-		// Tuple must end with comma, to mark it as a tuple.
-		if mode == modeTuple {
-			p.printf(",")
+	// In Default printing mode try to keep the original printing style
+	if tables.FormattingMode == tables.DefaultMode {
+		// If every element (including the brackets) ends on the same line where the next element starts,
+		// use the compact mode, otherwise use multiline mode
+		previousEnd := start.Line
+		for _, x := range *list {
+			start, end := x.Span()
+			if start.Line != previousEnd {
+				return false
+			}
+			previousEnd = end.Line
+		}
+		if end != nil && previousEnd != end.Pos.Line {
+			return false
+		}
+		return true
+	} else {
+		// In Build mode, use the forceMultiline and forceCompact values
+		if forceMultiLine {
+			return false
 		}
+		if forceCompact {
+			return true
+		}
+		// If neither of the flags are set, use compact mode only for empty or 1-element sequences
+		return len(*list) <= 1
+	}
+}
 
-	case forceCompact:
-		// Compact form but multiple elements.
-		for i, x := range list {
+// seq formats a list of values inside a given bracket pair (brack = "()", "[]", "{}").
+// The end node holds any trailing comments to be printed just before the
+// closing bracket.
+// The mode parameter specifies the sequence mode (see above).
+// If multiLine is true, seq avoids the compact form even
+// for 0- and 1-element sequences.
+func (p *printer) seq(brack string, start *Position, list *[]Expr, end *End, mode seqMode, forceCompact, forceMultiLine bool) {
+	isCompact := useCompactMode(start, list, end, forceCompact, forceMultiLine)
+	if mode != modeSeq {
+		p.printf("%s", brack[:1])
+	}
+	p.depth++
+
+	if isCompact {
+		for i, x := range *list {
 			if i > 0 {
 				p.printf(", ")
 			}
 			p.expr(x, precLow)
 		}
-
-	default:
+		// Single-element tuple must end with comma, to mark it as a tuple.
+		if len(*list) == 1 && mode == modeTuple {
+			p.printf(",")
+		}
+	} else {
```

Comment:
add `return`, remove `else`

Reply:
Can't return here, a closing bracket should be printed in the end of the function.

Reply:
you could return if you duplicate the code

p.depth--
if mode != modeSeq {
	p.printf("%s", brack[1:])
}

and you could de-dup by adding just after the p.depth++ above
defer func() {
  p.depth--
  if mode != modeSeq {
	p.printf("%s", brack[1:])
  }
}()

Reply:
Done.